### PR TITLE
fix: Add theme.d.ts to npm package

### PR DESCRIPTION
**Pull request by jameslnewell**
I was getting these errors:

```
ERROR in <project>/node_modules/treat/lib/commonjs/builder.d.ts(1,23):
TS2307: Cannot find module 'treat/theme'.
ERROR in <project>/node_modules/treat/lib/commonjs/types.d.ts(2,23):
TS2307: Cannot find module 'treat/theme'.
```

These errors appear to be caused by `theme.d.ts` in `packages/treat` not being published along with the rest of the package - you can verify `theme.d.ts` is missing from the published package on [unpkg](https://unpkg.com/browse/treat@1.0.3/).

I verified the code changes in this PR includes `theme.d.ts` by running `npm pack`.

I verified by manually copying `theme.d.ts` tp `node_modules/treat/theme.d.ts` that fixes the above errors.

### CODE REVIEW

**Comment by mattsjones** [comment]
Hey Bud. 👋 

This was originally excluded deliberately but we see that was probably the wrong choice. However, adding the file as it is currently will break consumers who override the `Theme` type as there will be two clashing definitions. We are working on a fix to solve this across the board. Hopefully won't be too long. 

A short term workaround is `"skipLibCheck": true` in your `tsconfig.json`.

**Comment by jameslnewell** [comment]
Ah I see. I wonder if using declaration merging like [the current recommendation for `styled-components`](https://www.styled-components.com/docs/api#typescript) would work?

**Comment by markdalgleish** [comment]
Yeah, that's actually what we we're using as a reference 👍

**Comment by jameslnewell** [comment]
> adding the file as it is currently will break consumers who override the Theme type as there will be two clashing definitions.

I just tested this, in my repo, after converting `Theme` from `any` to an interface (like `styled-components`) then I don't think it breaks anything for existing users???

If I copy the `treat/theme.d.ts` file across into `./node_modules/treat` then:
- it works with the default theme
- it works with the following theme definition

`./typings/treat/theme.d.ts`
```
declare module 'treat/theme' {
  export interface Theme {
    colors: {
      primary: string;
      secondary: string;
    }
  }
}
```

Then theme fns passed to `style(fn)` type the theme correctly.

**Comment by mattsjones** [comment]
Thanks for the update. Having a look at this solution now. By default, the theme should be type `any` not `{}` though so we might need to use a type util similar to [styled-components](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/ad1ea50852c22869ee9aeb2438f696e596e1e2e0/types/styled-components/index.d.ts#L353). 

Also, we should remove `theme.d.ts` from the `.npmignore` file instead of adding it to package.json.

File: packages/treat/theme.d.ts [hunk 1]
```diff
@@ -1,3 +1,7 @@
+type AnyIfEmpty<T extends object> = keyof T extends never ? any : T;
+
 declare module 'treat/theme' {
-  export type Theme = any;
+  export interface Theme {}
+  export type ThemeOrAny = AnyIfEmpty<Theme>
```
Comment (jameslnewell) [inline packages/treat/theme.d.ts]: is there another place you'd prefer to put this?

Comment (mattsjones) [reply packages/treat/theme.d.ts]: As long as it works for all scenarios I'm comfortable with it here. Just testing that this works across the board atm.

**Comment by mattsjones** [comment]
This is looking great. Just need to update the example code in the docs (`data-types.md`) to something like the following.

```diff
declare module 'treat/theme' {
+  type MyTheme = import('./types').Theme;

-  export type Theme = import('./types').Theme;
+  export interface Theme extends MyTheme {}
}
```

**Comment by jameslnewell** [comment]
Looks like the PR _could_ be a breaking change in that use case?

**Comment by mattsjones** [comment]
I agree, but seeing as the original solution was pretty much broken in a lot of situations, I see this more like fixing a broken implementation. Relying on `skipLibCheck` (which we had set for other reasons) is not a working solution. Also, this only affects the types of the library and should have very little impact on consumers.

**Pull request merged**